File: RotateArray/gcd.java
public class Solution { public void rotate(int[] nums, int k) { int temp = 0; int idxOld = 0;

    int aLen = nums.length;
    int i = 0;
    while (k >= aLen)
        k -= aLen;
    if (k == 0)
        return;

    int idxNew = k;    
    int GCD = 1;
    // get GCD.
    for (int j = k; j > 0; j--) {
        if ((aLen % j == 0) && (k % j == 0)) {
            GCD = j;
            break;
        }
    }
    System.out.print(GCD);
    while (i < aLen) {
        idxNew = (idxNew >= aLen ? idxNew-aLen : idxNew);
        temp = nums[idxNew];
        nums[idxNew] = nums[idxOld];
        nums[idxOld] = temp;
        i++;
        idxNew += k;
        if (i%(aLen/GCD) == 0) {
            idxNew ++;
            idxOld ++;
        }
    }
}
}

File: RotateArray/gcd.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int len = nums.size();
        int d = gcd(len, k=k%len);
        int pos, next_pos;
        int tmp, tmp_next;

        for (int i = 0; i < d; ++i) {
            pos = i;
            tmp = nums[pos];
            do {
                next_pos = (pos+k) % len;
                tmp_next = nums[next_pos];
                nums[next_pos] = tmp;

                pos = next_pos;
                tmp = tmp_next;
            } while (pos != i);
        }
    }

    int gcd(int a, int b)
    {
        int t;

        while (b != 0) {
            t = a%b;
            a = b;
            b = t;
        }
        return a;
    }
};
